src/odemis/driver/pmtctrl.py

```diff
@@ -420,18 +422,34 @@ def _sendCommand(self, cmd):
         """
         cmd (byte str): command to be sent to PMT Control unit.
         returns (byte str): answer received from the PMT Control unit
-        raises:
-            IOError: if an ERROR is returned by the PMT Control firmware.
         """
         cmd = cmd + b"\n"
         with self._ser_access:
             logging.debug("Sending command %s", to_str_escape(cmd))
-            self._serial.write(cmd)
+            try:
+                self._serial.write(cmd)
+            except IOError:
+                logging.warn("Failed to send command to PMT Control firmware, " +
+                             "trying to reconnect.")
+                if self._recovering:
+                    raise
+                else:
+                    self._tryRecover()
 
             ans = b''
             char = None
             while char != b'\n':
-                char = self._serial.read()
+                try:
+                    char = self._serial.read()
+                except IOError:
+                    logging.warn("Failed to read from PMT Control firmware, " +
+                                 "trying to reconnect.")
+                    if self._recovering:
+                        raise
+                    else:
+                        self._tryRecover()
+                        return
```

Comment:
Do you think this might work?
```
                else:
                    self._tryRecover()
                    # send command again
                    return self._sendCommand(cmd)
```

Reply:
The main things to consider is what happens if actually the device receives the command a second time, and what happens to the original caller if we raise an error.
In this driver, I think most of the cases sending the command twice in a row is very safe (just set a value to the same value), and the only place it could be a tiny bit dangerous would be when resetting the protection. My understanding is that this recovery part is already quite unlikely to happen (as it would more likely happen at the write before). So it should be fine to do.
Technically speaking, the RLock allows you to do that. You could imagine reaching some kind of recursion limit if it keeps calling itself, but that's probably the best thing to do in such bad case anyway.